api/controllers/reports.js

```diff
@@ -4,142 +4,203 @@ var _ = require('lodash');
 var async = require('async');
 var User = require('../../models/User');
 var Hours = require('../../models/Hours');
-//var Projects = require('../../models/Projects');
 var weeksSinceEpoch = require('../../lib/util').weeksSinceEpoch;
 
 /**
- * GET /report
- * Reportings dashboard.
+ * Gets hours for an epoch window (optionally), and returns them in flat layout.
  */
 
-/**
- * GET /dashboard
- * Hours dashboard.
- */
-exports.dashboard = function (req, res, next) {
+var getHours = function (epochSubtractor, callback) {
+
+    //Mapping function, returns flat layout.
+    function mapByProject(obj) {
+        var newArray = _.forEach(obj.projects, function (proj) {
+            proj.userId = obj.userId;
+            proj.week = obj.week;
+            return proj;
+        });
+        return newArray;
+    }
+
+    if (!epochSubtractor) {
+        epochSubtractor = 0;
+    }
+
     Hours.find({
         week: {
-            $gte: weeksSinceEpoch() - 3
+            $gte: weeksSinceEpoch() - epochSubtractor
         }
     }, function (err, docs) {
         if (err) {
-            return next(err);
+            return callback(err);
         }
-        var allData = _.groupBy(docs, 'userName');
-        var hours = [];
 
-        async.eachOfSeries(allData, function (value, key, cb) {
-            // look up related user to ensure they are active
-            User.findById(value[0].userId, function (err, user) {
-                if (err) {
-                    return cb(err);
-                } else if (user === null) {
-                    return cb();
-                } else if (user.inactive) {
-                    return cb();
-                }
-                var temp = {
-                    name: key,
-                    data: _.sortBy(value, 'week'),
-                    projects: []
-                };
-                _.forEach(value, function (val) {
-                    _.forEach(val.projects, function (project) {
-                        if (temp.projects.indexOf(project.name) < 0) {
-                            temp.projects.push(project.name);
-                        }
-                    });
-                });
-                hours.push(temp);
-                cb();
+        //Flatten all entries.
+        var mappedData = _.flatMap(docs, mapByProject);
+        callback(null, mappedData);
+
+        //lean() returns pojo.
+    }).lean();
```

Comment:
The objects returned from Mongoose wouldn't let me append anything to them (user information). Since it's for reporting, we don't actually need the getters/setters that come with them, so getting back a Pojo seemed like the best way to deal with it.

http://stackoverflow.com/questions/14504385/why-cant-you-modify-the-data-returned-by-a-mongoose-query-ex-findbyid